core/livepeernode.go

```diff
@@ -125,32 +134,143 @@ func (n *LivepeerNode) Start(ctx context.Context, bootIDs, bootAddrs []string) e
 
 	//Call eth client to create the job
 	blknum, err := n.Eth.LatestBlockNum()
 	if err != nil {
-		return ErrNotFound
+		return nil, ErrNotFound
 	}
-	tx, err := n.Eth.Job(strmID.String(), ethcommon.ToHex(transOpts)[2:], price, big.NewInt(0).Add(blknum, big.NewInt(DefaultJobLength)))
+
+	_, err = n.Eth.Job(strmID.String(), ethcommon.ToHex(transOpts)[2:], price, big.NewInt(0).Add(blknum, big.NewInt(DefaultJobLength)))
 	if err != nil {
 		glog.Errorf("Error creating transcode job: %v", err)
-		return err
+		return nil, err
 	}
 
-	err = n.Eth.CheckTx(tx)
+	job, err := n.Eth.WatchForJob(strmID.String())
+	if err != nil {
+		glog.Error("Unable to monitor for job ", err)
+		return nil, err
+	}
+	glog.V(common.DEBUG).Info("Got a new job from the blockchain: ", job.JobId)
+
+	glog.Infof("Created broadcast job. Id: %v Price: %v. Type: %v", job.JobId, price, ethcommon.ToHex(transOpts)[2:])
+
+	return job, nil
+}
+
+func (n *LivepeerNode) transcodeSegmentLoop(job *ethTypes.Job, segChan SegmentChan) error {
+	glog.V(common.DEBUG).Info("Starting transcode segment loop for ", job.StreamId)
+	cm, err := n.GetClaimManager(job)
 	if err != nil {
 		return err
 	}
+	resultStrmIDs := make([]StreamID, len(job.Profiles), len(job.Profiles))
+	broadcasters := make(map[StreamID]stream.Broadcaster)
+	sid := StreamID(job.StreamId)
+	for i, vp := range job.Profiles {
+		strmID, err := MakeStreamID(n.Identity, sid.GetVideoID(), vp.Name)
+		if err != nil {
+			glog.Error("Error making stream ID: ", err)
+			return err
+		}
+		resultStrmIDs[i] = strmID
+		broadcaster, err := n.VideoNetwork.GetBroadcaster(string(strmID))
+		if err != nil {
+			glog.Errorf("Error making new stream: %v", err)
+			return err
+		}
+		broadcasters[strmID] = broadcaster
+	}
+	tr := transcoder.NewFFMpegSegmentTranscoder(job.Profiles, n.WorkDir)
+	config := net.TranscodeConfig{
+		StrmID:              job.StreamId,
+		Profiles:            job.Profiles,
+		JobID:               job.JobId,
+		PerformOnchainClaim: cm != nil,
+	}
+	go func() {
+		for {
+			// XXX make context timeout configurable
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+			select {
+			case <-ctx.Done():
+				// timeout; clean up goroutine here
+				jid := job.JobId.Int64()
+				glog.V(common.DEBUG).Info("Segment loop timed out; closing ", jid)
+				n.segmentMutex.Lock()
+				if _, ok := n.SegmentChans[jid]; ok {
+					close(n.SegmentChans[jid])
+					delete(n.SegmentChans, jid)
+				}
+				n.segmentMutex.Unlock()
+				n.claimMutex.Lock()
+				if cm, ok := n.ClaimManagers[jid]; ok {
+					go func() {
+						err := cm.ClaimVerifyAndDistributeFees()
+						if err != nil {
+							glog.Errorf("Error claiming work for job %v: %v", jid, err)
+						}
+					}()
+					delete(n.ClaimManagers, jid)
+				}
+				n.claimMutex.Unlock()
+				return
+			case chanData := <-segChan:
+				chanData.res <- n.transcodeAndBroadcastSeg(&chanData.seg.Seg, chanData.seg.Sig, cm, tr, resultStrmIDs, broadcasters, config)
```

Comment:
We actually need to return something in order to un-block the channel so the broadcaster gets a response. A nil is good; that means the transcoding succeeded. Otherwise, the error gets serialized into a string and sent back in the response. (Yeah, serializing the error is a bit rough for now, but it's a big improvement over old approach of "blindly send the segment up the network and hope it gets transcoded.")